File: src/xyang/module.py
"""
YANG module representation.
"""

from __future__ import annotations

from typing import Dict, List, Optional, Set, TYPE_CHECKING
from dataclasses import dataclass, field

from .ast import YangStatementList

if TYPE_CHECKING:
    from .ast import YangIdentityStmt, YangStatement, YangTypedefStmt


@dataclass
class YangModule(YangStatementList):
    """Represents a YANG module. Statements list inherited from YangStatementList."""

    name: str = ""
    yang_version: str = "1.1"
    namespace: str = ""
    prefix: str = ""
    organization: str = ""
    contact: str = ""
    description: str = ""
    revisions: List[Dict[str, str]] = field(default_factory=list)
    # Set when this tree was parsed from a submodule (belongs-to argument).
    belongs_to_module: str = ""
    typedefs: Dict[str, 'YangTypedefStmt'] = field(default_factory=dict)
    identities: Dict[str, 'YangIdentityStmt'] = field(default_factory=dict)
    groupings: Dict[str, 'YangStatement'] = field(default_factory=dict)
    # Declared ``feature`` names in this module (RFC 7950 §7.18.1).
    features: Set[str] = field(default_factory=set)
    # Per-feature ``if-feature`` substatements (RFC 7950 §7.20.1.1); AND of expressions.
    feature_if_features: Dict[str, List[str]] = field(default_factory=dict)
    # import prefix (local) -> parsed module (RFC 7950 ``import``).
    import_prefixes: Dict[str, "YangModule"] = field(default_factory=dict)

    def own_prefix_stripped(self) -> str:
        return (self.prefix or "").strip("'\"")

    def resolve_prefixed_module(self, prefix: str) -> Optional["YangModule"]:
        """Module for ``prefix`` in this module's scope (own prefix or ``import``)."""
        if prefix == self.own_prefix_stripped():
            return self
        return self.import_prefixes.get(prefix)

    def get_typedef(self, name: str) -> Optional['YangTypedefStmt']:
        """Get a typedef by name."""
        return self.typedefs.get(name)
    
    def get_grouping(self, name: str) -> Optional['YangStatement']:
        """Get a grouping by name."""
        return self.groupings.get(name)

    def get_identity(self, name: str) -> Optional['YangIdentityStmt']:
        """Get an identity by name."""
        return self.identities.get(name)

File: ts/src/core/module.ts
import type {
  YangExtensionStmt,
  YangIdentityStmt,
  YangStatement,
  YangTypedefStmt
} from "./ast";
import { YangStatementList } from "./ast";

export class YangModule extends YangStatementList {
  name = "";
  yang_version = "1.1";
  namespace = "";
  prefix = "";
  organization = "";
  contact = "";
  description = "";
  revisions: Array<Record<string, string>> = [];
  belongs_to_module = "";
  typedefs: Record<string, YangTypedefStmt> = {};
  identities: Record<string, YangIdentityStmt> = {};
  groupings: Record<string, YangStatement> = {};
  features: Set<string> = new Set();
  feature_if_features: Record<string, string[]> = {};
  import_prefixes: Record<string, YangModule> = {};
  extensions: Record<string, YangExtensionStmt> = {};
  extension_runtime: Record<string, unknown> = {};

  constructor(init: Partial<YangModule> = {}) {
    super(init.statements ?? []);
    Object.assign(this, init);
  }

  own_prefix_stripped(): string {
    return (this.prefix || "").replace(/^['\"]|['\"]$/g, "");
  }

  ownPrefixStripped(): string {
    return this.own_prefix_stripped();
  }

  resolve_prefixed_module(prefix: string): YangModule | undefined {
    if (prefix === this.own_prefix_stripped()) {
      return this;
    }
    return this.import_prefixes[prefix];
  }

  resolvePrefixedModule(prefix: string): YangModule | undefined {
    return this.resolve_prefixed_module(prefix);
  }

  get_typedef(name: string): YangTypedefStmt | undefined {
    return this.typedefs[name];
  }

  getTypedef(name: string): YangTypedefStmt | undefined {
    return this.get_typedef(name);
  }

  get_grouping(name: string): YangStatement | undefined {
    return this.groupings[name];
  }

  getGrouping(name: string): YangStatement | undefined {
    return this.get_grouping(name);
  }

  get_identity(name: string): YangIdentityStmt | undefined {
    return this.identities[name];
  }

  getIdentity(name: string): YangIdentityStmt | undefined {
    return this.get_identity(name);
  }

  get_extension(name: string): YangExtensionStmt | undefined {
    return this.extensions[name];
  }

  getExtension(name: string): YangExtensionStmt | undefined {
    return this.get_extension(name);
  }
}
